playground/src/auto-importer/auto-importer.js

```diff
@@ -0,0 +1,245 @@
+// Copyright 2018 The AMPHTML Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CodeMirror from 'codemirror';
+
+const MISSING_ENGINE_PARAM = 'amphtml engine v0.js script';
+const AMP_BASE_URL = '"https://cdn.ampproject.org/v0.js"';
+const AMP_ENGINE_SCRIPT_TAG = `<script async src=${AMP_BASE_URL}></script>`;
+const COMPONENTS_URL = '/playground/amp-component-versions';
+
+let COMPONENT_VERSION_MAP = {};
+
+// A lookup to override what type of component an AMP component is
+const AMP_SCRIPT_TYPE_MAP = {
+  'amp-mustache': 'custom-template'
+};
+
+// Used as a lookup for attribute names that identify a script tag as being an
+// AMP component.
+const CUSTOM_ATTRS = {
+  'custom-element': 1,
+  'custom-template': 1
+};
+
+export function createAutoImporter(editor) {
+  return new AutoImporter(editor);
+}
+
+class AutoImporter {
+
+  constructor(editor) {
+    this.editor = editor;
+    this._fetchComponents();
+  }
+
+  update(validationResult) {
+    if (validationResult.status === 'FAIL') {
+      const missing = this._getMissingElements(validationResult);
+
+      if (Object.keys(missing.missingTags).length
+          || missing.missingBaseScriptTag) {
+        const existing = this._parseHeadTag();
+        // The action taken to insert any elements to fix the report of missing
+        // tags is determined by both a combination of looking at the list of
+        // missing elements and also the current state of the <head> tag.
+        this._insertMissingElements(missing, existing);
+      }
+    }
+  }
+
+  _fetchComponents() {
+    if (!Object.keys(COMPONENT_VERSION_MAP).length) {
+      let request = new Request(COMPONENTS_URL, {
+        headers: new Headers({'x-requested-by': 'playground'})
+      });
+      fetch(request)
+          .then((r) => r.json())
+          .then((data) => {
+            COMPONENT_VERSION_MAP = data;
+          });
+    }
+  }
+
+  /**
+   * Inserts missing <script> tags into the document, based on results from the
+   * AMP validator, and results from inspecting the current <head> structure.
+   *
+   * @param {!Object} missing The results from {@code _getMissingElements()}.
+   * @param {!Object} existing The results from {@code _parseHeadTag()}.
+   */
+  _insertMissingElements(missing, existing) {
+    const pos = existing.baseScriptTagEnd || existing.lastTag;
+    if (pos) {
```

Comment:
again: early exit `if(!pos) return` to avoid nesting

Reply:
Done